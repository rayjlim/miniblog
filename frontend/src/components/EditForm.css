.entry-bar {
  display: flex;
  flex-direction: row;
  padding: 0 1rem;
  justify-content: space-between;
}
.small {
  line-height: 1em;
  font-size: 0.8rem;
  margin: 0;
}

.delete-btn {
  margin-left: 'auto'
}
@media only screen and (max-width: 415px) {
  .well {
    flex-direction: column;
    gap: .5rem;
  }

  .btn {
    margin: 0 0;
  }

}
