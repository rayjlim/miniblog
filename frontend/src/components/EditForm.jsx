import React, { useState, useRef, useEffect } from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';
import DatePicker from 'react-date-picker';
import { format, parse } from 'date-fns';
import MarkdownDisplay from './MarkdownDisplay';
import constants from '../constants';

const EditForm = ({ entry, onSuccess }) => {
  const [date, setDate] = useState(
    parse(entry.date, constants.FULL_DATE_FORMAT, new Date()),
  );
  const escapedContent = entry.content.replace(
    /<br\s*\/>/g,
    `
`,
  );
  const [content, setContent] = useState(escapedContent);
  const textareaInput = useRef();

  function textChange() {
    const pattern = /@@([\w-]*)@@/g;
    const replacement = '<i class="fa fa-$1" ></i> ';
    console.log('textarea.value :>> ', textareaInput.current.value);
    textareaInput.current.value = textareaInput.current.value.replace(pattern, replacement);

    setContent(textareaInput.current.value);
  }

  /**
  * It takes the content and date from the form, and sends a PUT
  * request to the server with the updated
  * entry
  */
  // TODO: convert to customHook
  async function handleSave() {
    console.log('handleSave entry :', content, date);
    try {
      const token = window.localStorage.getItem(constants.STORAGE_KEY);
      const response = await fetch(
        `${constants.REST_ENDPOINT}/api/posts/${entry.id}`,
        {
          method: 'PUT',
          body: JSON.stringify({
            content,
            date: format(date, constants.FULL_DATE_FORMAT),
          }),
          mode: 'cors',
          cache: 'no-cache',
          credentials: 'same-origin',
          headers: {
            'Content-Type': 'application/json',
            'X-App-Token': token,
          },
          redirect: 'follow',
          referrerPolicy: 'no-referrer',
        },
      );

      console.log(response);
      onSuccess('Edit Done');
    } catch (error) {
      console.log(error);
      alert(error);
    }
  }

  // TODO: convert to customHook
  async function handleDelete() {
    const go = window.confirm('You sure?');
    if (!go) {
      return;
    }
    const { id } = entry;
    console.log(`handleDelete ${id}`);
    try {
      const token = window.localStorage.getItem(constants.STORAGE_KEY);
      const response = await fetch(
        `${constants.REST_ENDPOINT}/api/posts/${id}`,
        {
          method: 'DELETE',
          mode: 'cors',
          cache: 'no-cache',
          credentials: 'same-origin',
          headers: {
            'Content-Type': 'application/json',
            'X-App-Token': token,
          },
          redirect: 'follow',
          referrerPolicy: 'no-referrer',
        },
      );

      console.log(response);
      onSuccess('Delete Done');
    } catch (error) {
      console.log(error);
      alert(error);
    }
  }

  function dateChange(value = new Date()) {
    console.log('value :', value);
    if (value) {
      setDate(value);
    }
  }

  function checkKeyPressed(e) {
    console.log(`EditForm: handle key presss ${e.key}`);
    // console.log('131:' + markdown + ', hasChanges ' + hasChanges);
    if (e.altKey && e.key === 's') {
      console.log('S keybinding');
      // Note: this is a hack because the content value is taken from the init value
      document.getElementById('saveBtn').click();
    } else if (e.key === 'Escape') {
      document.getElementById('cancelBtn').click();
    }
  }
  /* A hook that is called when the component is mounted. */
  useEffect(() => {
    console.log('EditForm: useEffect');

    document.addEventListener('keydown', checkKeyPressed);
    return () => document.removeEventListener('keydown', checkKeyPressed);
  }, [entry]);

  return (
    <div className="well">
      {/* <button onClick={this.addFAtag} className="btn btn-info" style={templateStyle}>
                    fa-template
                </button> */}
      <h2>Edit Entry</h2>

      <div style={{ display: 'flex', flexDirection: 'row', padding: '0 1rem' }}>
        <div>
          <p style={{ lineHeight: '1em', fontSize: '.8rem', margin: '0' }}>use `@@fa-tag@@` for quick font-awesome icon</p>
          <p style={{ lineHeight: '1em', fontSize: '.8rem', margin: '0' }}>link: [link text](URL)</p>
          <a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links" target="_new">
            Cheatsheet
          </a>
          {', '}
          <a href="https://fontawesome.com/icons" target="_new">
            Font Awesome
          </a>
        </div>
        <div style={{ marginLeft: 'auto' }}>
          <button
            onClick={handleDelete}
            className="btn btn-danger pull-right"
            type="button"
          >
            <i className="fa fa-trash" />
            {' '}
            Delete
          </button>
        </div>
      </div>
      <div className="form-group">
        <textarea
          ref={textareaInput}
          onChange={() => textChange()}
          className="form-control"
          placeholder="Add ..."
          rows="8"
          defaultValue={content}
        />
      </div>
      <div className="editBtns">
        <button
          onClick={() => handleSave()}
          className="btn btn-primary"
          id="saveBtn"
          type="button"
        >
          <i className="fa fa-save" />
          {' '}
          Save
        </button>
        <DatePicker onChange={dateParam => dateChange(dateParam)} value={date} />
        <button
          onClick={() => onSuccess('')}
          className="btn btn-warning pull-right"
          id="cancelBtn"
          type="button"
        >
          <i className="fa fa-ban" />
          {' '}
          Cancel
        </button>
      </div>
      <div className="markdownDisplay preview dashBorder">
        <MarkdownDisplay source={content} />
      </div>
    </div>
  );
};

export default EditForm;

EditForm.propTypes = {
  entry: PropTypes.shape({
    id: PropTypes.string.isRequired,
    content: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  }).isRequired,
  onSuccess: PropTypes.func.isRequired,
};
