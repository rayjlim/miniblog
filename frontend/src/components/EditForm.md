EditForm example:

```js
let entry = {
  content: 'lorem ipsum',
  date: '2020-04-01',
  id: 1,
};
<EditForm entry={entry} onSuccess={() => console.log()} />;
```
