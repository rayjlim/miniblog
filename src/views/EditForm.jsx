import React from "react";
import moment from "moment";

class EditForm extends React.Component {
  constructor(props) {
    super(props);
    // create a ref to store the textInput DOM element

    this.save = this.save.bind(this);
    this.subToDate = this.subToDate.bind(this);
    this.addToDate = this.addToDate.bind(this);
  }
  render() {
    return (
      <div className="well">
        <h3>Edit Entry</h3>
        <div className="form-group">
          <textarea
            ref="content"
            onChange={event => event.preventDefault()}
            className="form-control"
            placeholder="Add ..."
            rows="8"
          >
            {this.props.entry.content}
          </textarea>
        </div>
        <div className="form-group">
          <input
            type="text"
            ref="date"
            onChange={event => event.preventDefault()}
            className="form-control"
            placeholder="Edit Date..."
            value={this.props.entry.date}
          />
        </div>
        <button onClick={this.subToDate} className="btn btn-info">
          subToDate
        </button>        
        <button onClick={this.addToDate} className="btn btn-success">
          addToDate
        </button>

        <button onClick={this.save} className="btn btn-primary">
          Save
        </button>

        <button
          onClick={this.props.clear}
          className="btn btn-warning pull-right"
        >
          Cancel
        </button>
        <button
          onClick={this.props.delete}
          className="btn btn-danger pull-right"
          id={this.props.entry.id}
        >
          Delete
        </button>
      </div>
    );
  }

  save ( e ) {
    const entry = {
      id: this.props.entry.id,
      content: this.refs.content.value.trim(),
      date: this.refs.date.value.trim()
    };
    this.props.submit(entry);
  }

  subToDate ( e ) {
    this.refs.date.value = moment(this.refs.date.value.trim()).subtract(1,'days').format('YYYY-MM-DD');
  }
  
  addToDate ( e ) {
    this.refs.date.value = moment(this.refs.date.value.trim()).add(1,'days').format('YYYY-MM-DD');
  }
}

export default EditForm;

