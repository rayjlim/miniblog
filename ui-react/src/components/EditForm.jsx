import React, { useState, useEffect } from 'react'; // eslint-disable-line no-unused-vars
import constants from '../constants';
import axios from 'axios';
import MarkdownDisplay from './MarkdownDisplay';

const EditForm = props => {
  let escapedContent = props.entry.content.replace(
    /<br\s*\/>/g,
    `
`
  );

  const [date, setDate] = useState(props.entry.date);
  const [content, setContent] = useState(escapedContent);

  let textareaInput = null;

  function textChange(text) {
    const pattern = /@@([\w-]*)@@/g;
    const replacement = '<i class="fa fa-$1" /> ';
    console.log('textarea.value :>> ', textareaInput.value);
    textareaInput.value = textareaInput.value.replace(pattern, replacement);

    setContent(textareaInput.value);
  }

  function handleSave() {
    const entry = {
      content,
      date, // TODO: check date format
    };
    console.log('axios entry :', entry);

    axios
      .put(`${constants.REST_ENDPOINT}api/posts/${props.entry.id}`, entry)
      .then(response => {
        console.log(response);
        props.onSuccess();
      })
      .catch(error => {
        console.log(error);
        alert(error);
      });
  }
  function handleClear() {
    props.onSuccess();
  }

  function handleDelete() {
    console.log('handleDelete ' + props.entry.id);
    axios
      .delete(`${constants.REST_ENDPOINT}api/posts/${props.entry.id}`)
      .then(response => {
        console.log(response);
        props.onSuccess();
      })
      .catch(error => {
        console.log(error);
        alert(error);
      });
  }

  return (
    <div className="well">
      {/* <button onClick={this.addFAtag} className="btn btn-info" style={templateStyle}>
                    fa-template
                </button> */}
      <h2>Edit Entry</h2>
      <span>use `@@fa-tag@@` for quick font-awesome icon</span>
      <p>
        link: [link text](URL){' '}
        <a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links">
          Cheatsheet
        </a>
      </p>

      <div className="form-group">
        <textarea
          ref={elem => (textareaInput = elem)}
          onChange={event => textChange(event.target.value)}
          className="form-control"
          placeholder="Add ..."
          rows="8"
          defaultValue={escapedContent}
        />
      </div>
      <div className="form-group">
        <input
          type="text"
          onChange={event => setDate(event.target.value)}
          className="form-control"
          placeholder="Edit Date..."
          defaultValue={props.entry.date}
        />
      </div>

      {/* <button onClick={this.subToDate} className="btn btn-info">
                    subToDate
                </button>
                <button onClick={this.addToDate} className="btn btn-success">
                    addToDate
                </button> */}
      <div className="editBtns">
        <button onClick={handleSave} className="btn btn-primary">
          <i className="fa fa-save" /> Save
        </button>
        <button onClick={handleClear} className="btn btn-warning pull-right">
          <i className="fa fa-ban" /> Cancel
        </button>
        <button onClick={handleDelete} className="btn btn-danger pull-right">
          <i className="fa fa-trash" /> Delete
        </button>
      </div>
      <div className="markdownDisplay">
        <MarkdownDisplay source={content} />
      </div>
    </div>
  );
};

export default EditForm;
