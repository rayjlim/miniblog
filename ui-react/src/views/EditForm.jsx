import React, { useState, useEffect } from 'react'; // eslint-disable-line no-unused-vars
import constants from '../constants';
import axios from 'axios';
import ReactMarkdown from 'react-markdown'; // eslint-disable-line no-unused-vars

const EditForm = (props) => {
	// class EditForm extends Component {
	//     constructor(props) {
	//         super(props);
	//         // create a ref to store the textInput DOM element

	//         this.save = this.save.bind(this);
	//         this.subToDate = this.subToDate.bind(this);
	//         this.addToDate = this.addToDate.bind(this);
	//         this.addFAtag = this.addFAtag.bind(this);
	//     }
	let escapedContent = props.entry.content.replace(
		/<br\s*\/>/g,
		`
`
	);
	const [ date, setDate ] = useState(props.entry.date);
	const [ content, setContent ] = useState(escapedContent);

	function handleSave() {
		const entry = {
			content,
			date // TODO: check date format
		};
		console.log('axios entry :', entry);
        const options = {
            method: 'PUT',
           };
        axios.put(`${constants.REST_ENDPOINT}api/posts/${props.entry.id}`,
        entry
        ,options)
			.then((response) => {
				console.log(response);
				// $('.toast').toast('dispose');
				// $('.toast-body').html('Updated The Entry');
				// $('.toast').toast({ delay: 4000 });
				// $('.toast').toast('show');
				props.onSuccess();
			})
			.catch((error) => {
				console.log(error);
				alert(error);
			});
	}
	function handleClear() {
		props.onSuccess();
	}

	let templateStyle = {
		float: 'right'
	};

	function handleDelete() {
		console.log('handleDelete ' + props.entry.id);
		axios
			.delete(`${constants.REST_ENDPOINT}api/posts/${props.entry.id}`)
			.then((response) => {
				console.log(response);
				// $('.toast').toast('dispose');
				// $('.toast-body').html('Removed Entry');
				// $('.toast').toast({ delay: 4000 });
				// $('.toast').toast('show');
				props.onSuccess();
			})
			.catch((error) => {
				console.log(error);
				// alert(error);
				props.onSuccess();
			});
	}

	return (
		<div className="well">
			{/* <button onClick={this.addFAtag} className="btn btn-info" style={templateStyle}>
                    fa-template
                </button> */}
			<strong>Edit Entry</strong>
			<p>
				link: [link text](URL){' '}
				<a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links">Cheatsheet</a>
			</p>
			{props.entry.id}
			<div className="form-group">
				<textarea
					onChange={(event) => setContent(event.target.value)}
					className="form-control"
					placeholder="Add ..."
					rows="8"
					defaultValue={escapedContent}
				/>
			</div>
			<div className="form-group">
				<input
					type="text"
					onChange={(event) => setDate(event.target.value)}
					className="form-control"
					placeholder="Edit Date..."
					defaultValue={props.entry.date}
				/>
			</div>

			{/* <button onClick={this.subToDate} className="btn btn-info">
                    subToDate
                </button>
                <button onClick={this.addToDate} className="btn btn-success">
                    addToDate
                </button> */}

			<button onClick={handleSave} className="btn btn-primary">
				Save
			</button>

			<button onClick={handleClear} className="btn btn-warning pull-right">
				Cancel
			</button>
			<button onClick={handleDelete} className="btn btn-danger pull-right">
				Delete
			</button>
            <ReactMarkdown source={content} escapeHtml={false} />

           
		</div>
	);
};

//     save(e) {
//         const entry = {
//             id: props.entry.id,
//             content: this.refs.content.value.trim(),
//             date: this.refs.date.value.trim()
//         };
//         props.submit(entry);
//     }

//     subToDate(e) {
//         this.refs.date.value = moment(this.refs.date.value.trim()).subtract(1, 'days').format('YYYY-MM-DD');
//     }

//     addToDate(e) {
//         this.refs.date.value = moment(this.refs.date.value.trim()).add(1, 'days').format('YYYY-MM-DD');
//     }

//     addFAtag(e) {
//         this.refs.content.value += `
// <i class="fas fa-" /> `;
//     }
// }

export default EditForm;
