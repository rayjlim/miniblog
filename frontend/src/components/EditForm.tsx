import { FunctionComponent } from 'react';

import MarkdownDisplay from './MarkdownDisplay';
import '../Types';
import useEditForm from '../hooks/useEditForm';

import './EditForm.css';

const EditForm: FunctionComponent<any> = ({ entry, onSuccess, }: { entry: EntryType | null, onSuccess: (msg: string, entry: EntryType) => void }) => {
  const escapedContent = entry?.content.replace(
    /<br\s*\/>/g,
    `
`,
  );

  const {
    textareaInput,
    content,
    dateInput,
    textChange,
    handleSave,
    handleDelete,
  } = useEditForm(entry, onSuccess);

  return (
    <div className="well">
      <h2>Edit Entry</h2>
      <div className="entry-bar">
        <div>
          <p className="small">use `@@fa-tag@@` for quick font-awesome icon</p>
          <p className="small">link: [link text](URL)</p>
          <a className="small" href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#links" target="_new">
            Cheatsheet
          </a>
          {', '}
          <a className="small" href="https://fontawesome.com/icons" target="_new">
            Font Awesome
          </a>
        </div>
        <button
          onClick={handleDelete}
          data-testid="deleteBtn"
          className="btn btn-danger pull-right delete-btn"
          type="button"
        >
          <i className="fa fa-trash" />
          {' '}
          Delete
        </button>
      </div>
      <div className="form-group">
        <textarea
          ref={textareaInput}
          onChange={() => textChange()}
          className="form-control"
          placeholder="Add ..."
          rows={8}
          defaultValue={escapedContent}
        />
      </div>
      <div className="editBtns">
        <button
          onClick={() => handleSave()}
          className="btn btn-primary"
          data-testid="saveBtn"
          id="saveBtn"
          type="button"
          title="alt + s"
        >
          <i className="fa fa-save" />
          {' '}
          Save
        </button>
        <input
          type="date"
          defaultValue={entry?.date || ''}
          ref={dateInput}

        />
        <button
          onClick={() => onSuccess('', { id: -1, content: '', date: '' })}
          className="btn btn-warning pull-right"
          data-testid="cancelBtn"
          id="cancelBtn"
          type="button"
          title="ESC"
        >
          <i className="fa fa-ban" />
          {' '}
          Cancel
        </button>
      </div>
      <div className="markdownDisplay preview dashBorder">
        <MarkdownDisplay source={content} />
      </div>
    </div>
  );
};

export default EditForm;
