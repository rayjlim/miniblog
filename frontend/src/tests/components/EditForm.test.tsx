// @ts-nocheck
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { vi } from 'vitest'
import jest from 'jest-mock';
import EditForm from "../../components/EditForm";

const fetchMock = vi.fn(() => ({
  json: vi.fn(() => Promise.resolve({ "id": "1", "content": "99 9 fasdf", "date": "2023-11-09", "user_id": "1" })),
  ok: true,
}));

vi.stubGlobal('fetch', fetchMock);

window.alert = () => { };
window.confirm = () => true;
console.log = () => {};

describe("EditForm component", () => {
  beforeEach(() => {
    HTMLCanvasElement.prototype.getContext = jest.fn();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  })

  it("should render EditForm component correctly", () => {

    render(<EditForm entry={{ id: 1, content: 'entry text', date: '2000-01-01' }} onSuccess={() => { }} />);
    const element = screen.getByText(/Cancel/);
    expect(element).not.toBeNull();
    expect(element).toBeInTheDocument();
  });

  it('should Cancel and responde', () => {

    const mockCallBack = jest.fn();

    render(<EditForm entry={{ id: 1, content: 'entry text', date: '2000-01-01', locations: '' }} onSuccess={mockCallBack} />);
    const btn = screen.getByTestId('cancelBtn');
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn);

    expect(mockCallBack).toHaveBeenCalledTimes(1);
    expect(mockCallBack.mock.calls[0][0]).toBe('');
  });

  it("should Save content on submit", async () => {
    const mockSuccessCb = jest.fn();

    render(<EditForm entry={{ id: 1, content: 'entry text', date: '2000-01-01', locations: '' }} onSuccess={mockSuccessCb} />);
    const btn = screen.getByTestId('saveBtn');
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn)
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch.mock.calls[0][0]).toContain("/api/posts/1");
    expect(fetch.mock.calls[0][1].method).toBe("PUT");
    expect(fetch.mock.calls[0][1].body).toMatchSnapshot();

    await waitFor(() => expect(mockSuccessCb).toHaveBeenCalledTimes(1));
    expect(mockSuccessCb.mock.calls[0][0]).toBe('Edit Done');

  });

  it("should Delete call and respond", async () => {
    const mockSuccessCb = jest.fn();

    render(<EditForm entry={{ id: 2, content: 'entry text', date: '2000-01-01' }} onSuccess={mockSuccessCb} />);
    const btn = screen.getByTestId('deleteBtn');
    expect(btn).toBeInTheDocument();
    fireEvent.click(btn)
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(fetch.mock.calls[0][0]).toContain("/api/posts/2")
    expect(fetch.mock.calls[0][1].method).toBe("DELETE")

    await waitFor(() => expect(mockSuccessCb).toHaveBeenCalledTimes(1));
    expect(mockSuccessCb.mock.calls[0][0]).toBe('Delete Done');

  });

  // test edit content
  // test edit date
  // test shortcut keys
  // test keychange font-awesome insert

});
