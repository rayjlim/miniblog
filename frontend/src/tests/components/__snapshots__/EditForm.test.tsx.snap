// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditForm component > should Save content on submit 1`] = `"{"id":1,"content":"entry text","date":"2000-01-01","locations":""}"`;
